Move CICD/ArgoCD out out Environments folder to "Config"

gaganhegde opened the pull request:
**What type of PR is this?**
 /kind design
![carbon](https://user-images.githubusercontent.com/48808456/83430370-903d1200-a403-11ea-95c6-d23d77ffac5f.png)

**Which issue(s) this PR fixes**:
https://issues.redhat.com/browse/GITOPS-147

Co Authored by :
@ishitasequeira

gaganhegde committed:
The changes saved locally

gaganhegde committed:
Made changes to add the ArgoCd and the Pipelines to the manifest

ishitasequeira committed:
corrected the init,imagerepo, bootstrap test cases

ishitasequeira committed:
corrected the test cases

gaganhegde committed:
Corrected the error in the trigger bindings

bigkevmcd commented on pkg/pipelines/argocd/argocd.go:
```diff
@@ -52,32 +51,38 @@ func Build(argoNS, repoURL string, m *config.Manifest) (res.Resources, error) {
 type argocdBuilder struct {
 	repoURL string
-	argoEnv *config.Environment
+	argoEnv *config.Argo
```
This should probably be renamed as it's no longer an `argoEnv` ?

gaganhegde replied:
Now changed to ArgoCDConfig and PipelinesConfig.

bigkevmcd commented on pkg/pipelines/argocd/argocd.go:
```diff
@@ -52,32 +51,38 @@ func Build(argoNS, repoURL string, m *config.Manifest) (res.Resources, error) {
 
 func (b *argocdBuilder) Application(env *config.Environment, app *config.Application) error {
-	basePath := filepath.Join(config.PathForEnvironment(b.argoEnv), "config")
+	// basePath := filepath.Join(config.PathForArgoEnvironment(b.argoEnv), "config")
```
This line is not doing anything ?

bigkevmcd commented on pkg/pipelines/argocd/argocd.go:
```diff
@@ -52,32 +51,38 @@ func Build(argoNS, repoURL string, m *config.Manifest) (res.Resources, error) {
 		return nil
 	}
-	basePath := filepath.Join(config.PathForEnvironment(b.argoEnv), "config")
+	// basePath := filepath.Join(config.PathForArgoEnvironment(env), "config")
```
This line can come out too.

bigkevmcd commented on pkg/pipelines/argocd/argocd.go:
```diff
@@ -38,9 +38,8 @@ func Build(argoNS, repoURL string, m *config.Manifest) (res.Resources, error) {
-	argoEnv, err := m.GetArgoCDEnvironment()
-	// If there's no ArgoCD environment, then we don't need to do anything.
-	if err != nil {
+	argoEnv, err := m.GetArgo()
```
You're not using `err` here, but you can remove this, as it now returns `nil` if it doesn't exist, and there can't be any more than one now.

bigkevmcd commented on pkg/pipelines/argocd/argocd.go:
```diff
@@ -52,32 +51,38 @@ func Build(argoNS, repoURL string, m *config.Manifest) (res.Resources, error) {
 	filename := filepath.Join(basePath, "kustomization.yaml")
 	resourceNames := []string{}
-	for k, _ := range b.files {
+	for k, _ := range files {
```
You can simplify this to `k := range files {`

bigkevmcd commented on pkg/pipelines/config/config.go:
```diff
@@ -71,37 +81,23 @@ func (m *Manifest) AddService(envName, appName string, svc *Service) error {
-	for _, env := range m.Environments {
-		if env.IsCICD {
-			envs = append(envs, env)
+func (m *Manifest) GetCICD() (*Cicd, error) {
```
Couple of things here...

I'd rename `Cicd` to `CICDConfig`, or possibly `PipelinesConfig`.

It doesn't need to return an error, if you look at the code, it never returns a non-nil case, but you don't need to, as that was for when you could have multiple environments.

bigkevmcd commented on pkg/pipelines/config/config.go:
```diff
@@ -71,37 +81,23 @@ func (m *Manifest) AddService(envName, appName string, svc *Service) error {
-		if env.IsArgoCD {
-			envs = append(envs, env)
+// GetArgo returns the ArgoCD Environment if one exists.
+func (m *Manifest) GetArgo() (*Argo, error) {
```
See above for the CICD case...no error, and maybe rename to `ArgoCDConfig`?

bigkevmcd commented on pkg/pipelines/config/config_test.go:
```diff
@@ -76,78 +84,78 @@ func TestManifestWalk(t *testing.T) {
-			},
-			{
-				Name: "development",
+// func TestManifestWalkCallsCICDEnvironmentLast(t *testing.T) {
```
This can be removed?

bigkevmcd commented on pkg/pipelines/config/config_test.go:
```diff
@@ -76,78 +84,78 @@ func TestManifestWalk(t *testing.T) {
+// 	}
+// }
 
 func TestEnviromentSorting(t *testing.T) {
```
You can now simplify this entirely, we only need one case, `testEnv` can go away.

bigkevmcd commented on pkg/pipelines/config/config_test.go:
```diff
@@ -179,48 +187,69 @@ func TestEnviromentSorting(t *testing.T) {
-		names []testEnv
-		want  string
-		err   string
+		// names []testEnv
```
These lines can be removed?

bigkevmcd commented on pkg/pipelines/config/config_test.go:
```diff
@@ -179,48 +187,69 @@ func TestEnviromentSorting(t *testing.T) {
-		{[]testEnv{{"prod", false, false}, {"staging", false, false}, {"dev", false, false}}, "", ""},
-		{[]testEnv{{"test-cicd", true, false}, {"staging", false, false}, {"dev", false, false}}, "test-cicd", ""},
-		{[]testEnv{{"test-cicd", true, false}, {"oc-cicd", true, false}, {"dev", false, false}}, "", "found multiple CI/CD environments"},
+		// {[]testEnv{{"prod", false, false}, {"staging", false, false}, {"dev", false, false}}, "", ""},
```
Again, see above.

bigkevmcd commented on pkg/pipelines/config/config_test.go:
```diff
@@ -179,48 +187,69 @@ func TestEnviromentSorting(t *testing.T) {
 		t.Run(fmt.Sprintf("test %d", i), func(rt *testing.T) {
-			m := &Manifest{Environments: makeEnvs(tt.names)}
-			env, err := m.GetCICDEnvironment()
+			// m := &Manifest{Environments: makeEnvs(tt.names)}
```
Some more lines that can come out?

bigkevmcd commented on pkg/pipelines/config/config_test.go:
```diff
@@ -179,48 +187,69 @@ func TestEnviromentSorting(t *testing.T) {
-			if tt.want != "" && (env.Name != tt.want) {
-				rt.Errorf("found incorrect CICD environment, got %s, want %s", env.Name, tt.want)
-			}
+			// if tt.want != "" && (env.Name != tt.want) {
```
And here too...

bigkevmcd commented on pkg/pipelines/config/config_test.go:
```diff
@@ -179,48 +187,69 @@ func TestEnviromentSorting(t *testing.T) {
-	if env.Name != "prod" {
-		t.Fatalf("got the wrong environment back: %#v", env)
-	}
+// func TestGetEnvironment(t *testing.T) {
```
We can remove this?

bigkevmcd commented on pkg/pipelines/config/validate.go:
```diff
@@ -61,9 +76,6 @@ func (vv *validateVisitor) Environment(env *Environment) error {
 func (vv *validateVisitor) Application(env *Environment, app *Application) error {
 	appPath := yamlPath(PathForApplication(env, app))
-	if env.IsSpecial() {
-		vv.errs = append(vv.errs, invalidEnvironment(env.Name, "A special environment cannot contain applications.", []string{appPath}))
```
It looks like you're not using `invalidEnvironment` anywhere now?

bigkevmcd commented on pkg/pipelines/config/validate.go:
```diff
@@ -24,6 +26,7 @@ func (m *Manifest) Validate() error {
 		serviceURLs:  map[string][]string{},
+		manifest:     m,
 	}
 	m.Walk(vv)
```
m.Walk can return an error, and it's not being picked up here.

bigkevmcd commented on pkg/pipelines/config/config.go:
```diff
@@ -116,21 +112,29 @@ type Environment struct {
 	Apps      []*Application `json:"apps,omitempty"`
 	// TODO: this should check that there is 0 or 1 CICD environment in the
```
This doesn't seem to be true anymore?

bigkevmcd replied:
Also, the `Environment` comment refers to the CICD environments.

bigkevmcd commented on pkg/pipelines/config/config.go:
```diff
@@ -116,21 +112,29 @@ type Environment struct {
+	Argo *Argo `json:"Argo,omitempty"`
+}
+
+//Cicd checks for the cicd environments
```
This comment doesn't seem to be right?

bigkevmcd commented on pkg/pipelines/config/config.go:
```diff
@@ -116,21 +112,29 @@ type Environment struct {
+	Namespace string `json:"namespace,omitempty"`
+}
+
+//Argo checks for the cicd environments
```
Not sure this one is right either?

wtam2018 commented on pkg/pipelines/config/config_test.go:
```diff
@@ -76,79 +84,6 @@ func TestManifestWalk(t *testing.T) {
 	}
 }
 
-func TestManifestWalkCallsCICDEnvironmentLast(t *testing.T) {
```
Will this test be replaced by a new one?

gaganhegde replied:
This wont be necessary as the the cicd environment is now present in a separate  struct, so the sorting is not necessary.

wtam2018 commented on pkg/pipelines/config/validate.go:
```diff
@@ -46,6 +48,18 @@ func (vv *validateVisitor) validateServiceURLs() []error {
+				vv.errs = append(vv.errs, errors.New("Cannot add a new application to Config environments"))
+			}
+		}
+	}
```
This check you have here is checking for namespace collision between any plain old Environment and "specail" CICD/ArgoCD which is a good validation.   However, we got here not because we are adding any applications so the error messages are  off.    Also, we should use the error functions like other errors which log the env.Path (otherwise, good luck troubleshooting what went wrong with the manifest file.)

You should also add unit tests for these cases.

Instead of keeping manifest in vw, I would keep the namespaces of "config" (i.e. CICD and ArgoCD namespaces).  Then, your checking code here wouldn't need to reference the config namespaces from manifest.   It will make the code looks cleaner.   It means that you should walk the Config in config.walk() which is another problem.   You need to validate the names of CICD and ArgoCD environments which have been skipped now.

You should add unit tests for these validation.

wtam2018 commented on pkg/pipelines/environments/environments.go:
```diff
@@ -62,15 +62,15 @@ func (b *envBuilder) Service(env *config.Environment, svc *config.Service) error
-	}
+	// if env.IsSpecial() {
+	// 	return nil
+	// }
```
Please delete.

wtam2018 commented on pkg/pipelines/argocd/argocd.go:
```diff
@@ -38,46 +38,49 @@ func Build(argoNS, repoURL string, m *config.Manifest) (res.Resources, error) {
 	filename := filepath.Join(basePath, env.Name+"-"+app.Name+"-app.yaml")
 	argoFiles[filename] = makeApplication(env.Name+"-"+app.Name, b.argoNS, defaultProject, env.Name, defaultServer, makeSource(env, app, b.repoURL))
 	b.files = res.Merge(argoFiles, b.files)
+	err := argoEnvironmentResources(b.argoEnv, b.files)
```
It's argocd.  Or you mean https://en.wikipedia.org/wiki/Argo_(2012_film)

gaganhegde replied:
Now changed to ArgoCDConfig and PipelineConfig

wtam2018 commented on pkg/pipelines/config/config.go:
```diff
@@ -116,21 +112,29 @@ type Environment struct {
+type Config struct {
+	CICD *Cicd `json:"CICD,omitempty"`
+	Argo *Argo `json:"Argo,omitempty"`
+}
```
Argo or ArgoCD?

wtam2018 commented on pkg/pipelines/config/config.go:
```diff
@@ -22,10 +21,21 @@ func PathForEnvironment(env *Environment) string {
 }
 
+//PathForCICDEnvironment returns the path only for the CICD environment
+func PathForCICDEnvironment(cicd *Cicd) string {
```
If CICD and ArgoCD are no longer Environments, we should probably stop calling them "Environments".

gaganhegde replied:
Changed to PathForArgoCdConfig() and PathForCICDConfig().

wtam2018 commented on pkg/pipelines/service_test.go:
```diff
@@ -348,57 +361,156 @@ func TestServiceWithArgoCD(t *testing.T) {
-						Name: "test-app",
-						ServiceRefs: []string{
-							"test-svc",
+	if !withCICD && withArgoCD {
```
Please clean this up.   It's pretty messy.

wtam2018 commented on pkg/pipelines/tekton_builder.go:
```diff
@@ -63,14 +64,17 @@ func (tk *tektonBuilder) Service(env *config.Environment, svc *config.Service) e
-	if env.IsCICD {
-		triggers, err := createTriggersForCICD(tk.gitOpsRepo, env)
+	// if env.IsCICD {
+	// 	triggers, err := createTriggersForCICD(tk.gitOpsRepo, env)
```
Please delete.

wtam2018 commented on pkg/pipelines/config/config.go:
```diff
@@ -116,21 +112,29 @@ type Environment struct {
+
+//Cicd checks for the cicd environments
+type Cicd struct {
+	Namespace string `json:"namespace,omitempty"`
```
We should call it "Name" like any entities in the manifest.   We just create namespaces for them but they are names of something (like Env, App).  In this case, it is a Config.

gaganhegde replied:
Now changed to Name.

gaganhegde replied:
Now Changed to Name.

bigkevmcd replied:
I'm not so sure about this.

ArgoCD it should at least be the namespace, as it's the namespace that the ArgoCD Applications need to go into (which is the ArgoCD namespace).

The CICD one is perhaps different, but at least for argo, we want folks to realise that it's the _namespace_.

wtam2018 commented on pkg/pipelines/config/config.go:
```diff
@@ -116,21 +112,29 @@ type Environment struct {
+
+//Argo checks for the cicd environments
+type Argo struct {
+	Namespace string `json:"namespace,omitempty"`
```
We should call it "Name" like any entities in the manifest.   We just create namespaces for them but they are names of something (like Env, App).  In this case, it is a Config.

wtam2018 commented on pkg/pipelines/environments/environments_test.go:
```diff
@@ -135,36 +143,64 @@ func filesFromResources(r res.Resources) []string {
+		}
 	}
-	return cfg
+	return nil
```
The caller downstream does not handle nil.   This will cause the caller to crash.

wtam2018 reviewed (changes requested):
I would like to have another look after all issues have been resolved.

gaganhegde committed:
Resolved some of the comments

ishitasequeira committed:
fixed service_test.go

ishitasequeira committed:
refactored the code and corrected the test cases

gaganhegde committed:
Included the validation for the config directory

ishitasequeira committed:
refactored the code and modularized the test cases

wtam2018 committed:
Merge branch 'bootstrap' of https://github.com/rhd-gitops-example/odo into directory-update

wtam2018 committed:
Resolved merge conflicts.

ishitasequeira committed:
fixed the error yaml files

wtam2018 commented on pkg/pipelines/bootstrap.go:
```diff
@@ -166,31 +166,40 @@ func bootstrapServiceDeployment(dev *config.Environment) (res.Resources, error)
 			env.Pipelines = defaultPipelines(repo)
+			envs = append(envs, env)
+		}
+		if k == "stage" {
```
Why are we adding this "stage" Environemnt in this PR?

wtam2018 replied:
@gaganeggday @ishitasequeira   Any thoughts?

wtam2018 commented on pkg/pipelines/config/config.go:
```diff
@@ -70,67 +82,61 @@ func (m *Manifest) AddService(envName, appName string, svc *Service) error {
+// The PipelineConfig should not have any applications defined.
+type Config struct {
+	PipelineConfig *Pipeline `json:"PipelineConfig,omitempty"`
+	ArgoCDConfig   *ArgoCD   `json:"ArgoCDConfig,omitempty"`
```
lower caases json declaration.

wtam2018 commented on pkg/pipelines/config/config.go:
```diff
@@ -22,10 +21,23 @@ func PathForEnvironment(env *Environment) string {
+
+// func PathForArgoCDConfig(argo *Argo) string {
+// 	return filepath.Join("config", argo.Namespace)
+// }
```
Please delete.

wtam2018 committed:
Fix yaml marshalling problem in TestAddService()

bigkevmcd commented on pkg/pipelines/config/config.go:
```diff
@@ -70,67 +82,61 @@ func (m *Manifest) AddService(envName, appName string, svc *Service) error {
-	IsArgoCD bool `json:"argo,omitempty"`
+}
+
+//Config are the Config environments that constitute the argocd and pipeline environment
```
This could be tidied up a bit...

Space before the `Config`.

Move the `PipelineConfig` description down to above the relevant struct field.

Add a similar description for the `ArgoCDConfig`

bigkevmcd committed:
Fix the config directory.

bigkevmcd committed:
Fix the ArgoCD.

bigkevmcd committed:
Fix environments.

bigkevmcd committed:
Fix imagerepo.

bigkevmcd committed:
Fix up the webhook package.

bigkevmcd committed:
Complete the rework.

wtam2018 committed:
* Fix a func name
* Remove unused test files.

wtam2018 commented:
I think this PR should be merged together PR to change kustomization resource referances and  apply script.   If we don't merge together, we will give a broken system.   People can't try/play with this baby.   Perhaps, the two other PRs can be merged to a branch to run end-to-end tests before merging (together) to bootstrap?

bigkevmcd committed:
Fix duplication of eventlistener triggers.

wtam2018 committed:
Merge branch 'bootstrap' of https://github.com/rhd-gitops-example/odo into directory-update

wtam2018 committed:
* Add back TestManifestWallCalls
* Remove dead fields: cicd and argocd in config_test.go

wtam2018 committed:
* add environmentNameCollidesWithConfig() and include path in the error.
* log "path" when validName() fails  on config.argocd and config.pipelines
* remove dead error func: invalidEnvironment
* add validate test case for invalid config.pipeline name.
* add validate test case for config.pipeline name and environment name collision.

wtam2018 committed:
Use generic invalidEnvironment() instead of environmentNameCollidesWithConfig()

wtam2018 committed:
Minor error message fix.

wtam2018 committed:
* rename variables cicd -> pipelinesConfig
* fix lint warnings in tests
* add  missing error logging in environment_test.go
* remove dead const in environment.go

wtam2018 committed:
* code clenaup: delete dead code, fix lint warnings
* simplify buildManifest() in servcie_test.go.   It should not return nil.

wtam2018 committed:
* Fix a bug in bootstrapEnvironments() that looks for for "stage"  environment.   If there is is a "foo" environment in namespace map, "foo" will not be bootstrapped.

wtam2018 reviewed (changes requested):
Don't merge until the following is ready.
* Dry-run script
* Set Kustomization references.

chetan-rns committed:
Generate the triggers for CI/CD env separately (#114)

Fixed a bug that generates triggers multiple times by generating triggers outside of "Environment" walk function which can get called when each environment is "walked"

chetan-rns committed:
Add a script to dry-run across all environments/applications (#112)

This commit adds a script to the task so that it can apply/dry-run across all the environments/applications. It makes use of a template to inject the parameters.
The script will selectively apply the resources based on the presence/absence of ArgoCD.

wtam2018 committed:
Add support for linking apps to environments and the reverse. (#115)

This adds support for determining whether or not an environment is linked to
its constituent apps, or linking apps to their environments.

The former is necessary to be able to bring up all the apps in an environment,
and the latter is necessary to bring up apps with their dependencies in the
environment.

Co-authored-by: Kevin McDermott <bigkevmcd@gmail.com>

wtam2018 committed:
Removed debug  log statements

wtam2018 committed:
Add back TestEnviromentSorting()

wtam2018 committed:
Remove TestEnvironmentSorting() which is no longer needed.    That was necessary before for testing purposes, because we had to ensure that some environments (argocd/cicd) were sorted to the end

wtam2018 committed:
Fix a bug ithat kuctomization.yaml is included in argocd/config/kustomization.yaml's resources.  The problem was that argoCDConfigResources() was called in the Application's walk() function.   Creation of argcd config resources (kustomization.yaml) was called as many times as the number of Applications.
